File: aws/lambda-hmac.js
const crypto = require('crypto');

exports.handler = async (event) => {

  const request_signature = event.headers ? event.headers['x-header-signature'] : "";
  const computed_signature = crypto.createHmac("sha256", process.env.SHARED_SECRET)
                                .update(Buffer.from(event.body, "base64").toString('utf8'))
                                .digest("base64");

  // TODO: is there a constant-time comparison method in the crypto lib?
  if (request_signatue !== computed_signature) {
    // auth failed!
  }

File: aws/lambda-hmac.rb
def lambda_handler(event:, context:)
  provided_signature = event['headers']['x-request-signature'].split('=').last
  generated_signature = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha1'), ENV['WEBHOOK_SECRET'], event['body'])

  # needs a constant-time comparison method to prevent timing attacks
  if provided_signature != generated_signature
    raise "signature mismatch :("


